tuf/data/root.go

```diff
@@ -23,14 +24,53 @@ type Root struct {
 	ConsistentSnapshot bool                 `json:"consistent_snapshot"`
 }
 
+// isValidRootStructure returns an error, or nil, depending on whether the content of the struct
+// is valid for root metadata.  This does not check signatures or expiry, just that
+// the metadata content is valid.
+func isValidRootStructure(r Root) error {
+	expectedType := TUFTypes[CanonicalRootRole]
+	if r.Type != expectedType {
+		return ErrInvalidMeta{
+			Role: CanonicalRootRole, Msg: fmt.Sprintf("expected type %s, not %s", expectedType, r.Type)}
+	}
+	if len(r.Roles) < 4 {
+		return ErrInvalidMeta{Role: CanonicalRootRole, Msg: "does not have all required roles"}
+	} else if len(r.Roles) > 4 {
+		return ErrInvalidMeta{Role: CanonicalRootRole, Msg: "specifies too many roles"}
+	}
+
+	for _, roleName := range BaseRoles {
+		roleObj, ok := r.Roles[roleName]
+		if !ok || roleObj == nil {
```

Comment:
I think we'd have to check if the role is nil, because that's a special case error for root roles, but I will factor out the rest of the key checks into a helper function.